design/thrift-proxy-design.md

```diff
@@ -0,0 +1,215 @@
+This functionality is particularly valuable for organizations that rely on the Thrift Protocol as their primary RPC implementation.
+
+Key features of the Envoy Thrift Proxy include:
+
+* Route Matching: The Thrift Proxy enables granular route matching based on various criteria, such as the Thrift method name, Thrift service and name, and a set of headers. 
+This level of specificity ensures accurate routing of Thrift requests to their intended destinations.
+* Rate Limiting: It supports rate limit filtering within the Thrift filter chain.
+* Cluster Management: The Thrift Proxy offers support for weighted clusters and mirrored clusters, enabling efficient load balancing and fail-over strategies.
+* Route Discovery: It facilitates route discovery mechanisms, making it easier to dynamically configure and manage Thrift routing rules.
+
+### Contour's Current Capabilities
+
+As of the current Contour release, the project provides "TCP Proxying" functionality. 
+While this is valuable, it does not include native support for Thrift Proxying.
+
+### The Need for Thrift Proxy in Contour
+
+The absence of Thrift Proxy support in Contour creates limitations for organizations that heavily rely on the Thrift Protocol. 
+As the Thrift Protocol continues to be a popular choice for inter-service communication, integrating Thrift Proxy into Contour is a logical step forward. 
+This integration will address the following key needs:
+* Protocol Support: Adding Thrift Proxy support ensures that Contour can seamlessly route and manage Thrift-based applications alongside its existing capabilities for HTTP and gRPC traffic.
+* Enhanced Routing Control: With Thrift Proxy, Contour gains the ability to implement fine-grained routing logic for Thrift requests, enabling intelligent routing decisions based on Thrift-specific criteria.
+* Interoperability: Thrift Proxy ensures that Contour can easily integrate with systems and services utilizing the Thrift Protocol.
+
+By introducing Thrift Proxy support into Contour, we aim to elevate Contour's versatility, making it a more comprehensive solution for modern, distributed application architectures.
+
+## Goals
+
+* Implement an equivalent to HTTPProxy, called ThriftProxy, which allows users to define their Thrift routing logic and services: This goal aims to provide users with a dedicated resource, ThriftProxy, 
+where they can specify custom Thrift routing rules, enabling fine-grained control over how Thrift traffic is handled within Contour.
+* Add a new Listener for the Thrift protocol: This goal involves enhancing Contour's capabilities by introducing a new Listener specifically designed to handle Thrift protocol traffic.
+* Implement ThriftRoute discovery: ThriftRoute discovery mechanisms will allow for dynamic configuration and management of Thrift routing rules.
+
+## Non Goals
```

Comment:
Would add Thrift over HTTP is a non-goal, since you can use existing HTTP routing capabilities of Contour